cloud-controller-manager/do/firewall_controller.go

```diff
@@ -0,0 +1,279 @@
+			return nil, fmt.Errorf("failed to retrieve firewall by ID: %s", err)
+		}
+		if resp != nil && resp.StatusCode == 404 {
+			klog.Infof("unable to retrieve firewall by ID because it no longer exists")
+		}
+		if fw != nil {
+			return fw, nil
+		}
+	}
+
+	// iterate through firewall API provided list and return the firewall with the matching firewall name.
+	f := func(fw godo.Firewall) bool {
+		return cmp.Equal(fw.Name, fm.workerFirewallName)
+	}
+	fw, err := filterFirewallList(ctx, fm.client, f)
+	if err != nil {
+		return nil, fmt.Errorf("failed to retrieve list of firewalls from DO API: %s", err)
+	}
+	return fw, nil
+}
+
+// Set applies the given inbound rules to the CCM worker firewall when the current rules and target rules differ.
+func (fm *firewallManagerOp) Set(ctx context.Context, svcInboundRules []godo.InboundRule) error {
+	fm.fwCache.mu.RLock()
+	targetFirewall := fm.fwCache.firewall
+	fm.fwCache.mu.RUnlock()
+
+	fwSvc := fm.client.Firewalls
+
+	fm.fwCache.mu.RLock()
+	isEqual := cmp.Equal(targetFirewall.InboundRules, svcInboundRules)
+	fm.fwCache.mu.RUnlock()
+
+	if !isEqual {
+		currentFirewall, err := fm.Get(ctx)
+		if err != nil {
+			// Log error then create firewall and update cache, since firewall does not exist.
+			klog.Errorf("failed to get firewall state: %s", err)
+			currentFirewall, err = fm.createFirewall(ctx, fwSvc, svcInboundRules)
+			if err != nil {
+				return err
+			}
+			fm.updateCache(currentFirewall)
+			return nil
+		} else {
+			// Update api firewall inbound rules with the expected inbound rules then update cache.
+			fr := &godo.FirewallRequest{
+				Name:         fm.workerFirewallName,
+				InboundRules: svcInboundRules,
+			}
+			currentFirewall, resp, err := fwSvc.Update(ctx, currentFirewall.ID, fr)
+			if err != nil {
+				if resp == nil || resp.StatusCode != http.StatusNotFound {
+					return fmt.Errorf("failed to update firewall: %v", err)
+				}
+				// firewall does not exist, so we should create a new one.
+				currentFirewall, err = fm.createFirewall(ctx, fwSvc, svcInboundRules)
+				if err != nil {
+					return fmt.Errorf("failed to add new firewall inbound rules: %s", err)
+				}
+			}
+			// if the firewall api update has succeeded, proceed to update the local firewall
+			// cache.
+			fm.updateCache(currentFirewall)
+		}
+	}
+	return nil
+}
+
+func (fm *firewallManagerOp) createFirewall(ctx context.Context, fwSvc godo.FirewallsService, svcInboundRules []godo.InboundRule) (*godo.Firewall, error) {
+	fr := &godo.FirewallRequest{
+		Name:         fm.workerFirewallName,
+		InboundRules: svcInboundRules,
+		// Tags:          currentFirewall.Tags,
+	}
+	currentFirewall, _, err := fwSvc.Create(ctx, fr)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create new firewall: %v", err)
+	}
+
+	return currentFirewall, nil
+}
+
+func (fm *firewallManagerOp) updateCache(currentFirewall *godo.Firewall) {
+	fm.fwCache.mu.RLock()
```

Comment:
This must be a `Lock` (not `RLock`) because we intend to write.

Similar for the unlocking part.